fix(python) fix `or` conflicts with string highlighting

MohamedAli00949 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
I have fixed the error of highlighting the reserved words such as 'or' in the below code.
```
print(test()or'string')
```
That should be in the below image.
![image](https://github.com/highlightjs/highlight.js/assets/76850143/06f4b5c3-18bc-4f2f-9d45-9a2ad7959adf)
Not as this highlight.
![image](https://github.com/highlightjs/highlight.js/assets/76850143/5805a432-1f3e-4b7b-be6c-fbb39b4bc267)

That happens because of the string prefixes (f, r, u, and b) or (F, R, U, and B).

This issue has been described at #3798

### Changes
I have found that happens from string variants begin RegExes with the prefix `(\W|^)`.

### Checklist
- [x] Update the string variants begin RegExes with the prefix `(\W|^)`.

joshgoebel commented on src/languages/python.js:
```diff
@@ -176,7 +176,7 @@ export default function(hljs) {
     variants: [
       {
-        begin: /([uU]|[bB]|[rR]|[bB][rR]|[rR][bB])?'''/,
+        begin: /(\W|^)([uU]|[bB]|[rR]|[bB][rR]|[rR][bB])?'''/,
```
Aren't you really just expressing "word boundary" here... and can't that be done much simpler using `\b`?

MohamedAli00949 replied:
I think it's doing the same thing here. 
But it didn't come first to my mind. 
Thanks for your advice and I will replace `(\W|^)` with it.

MohamedAli00949 replied:
@joshgoebel please take another look at my changes.
Many thanks for considering my request.

joshgoebel commented:
We need to add some `markup` tests as well to prove this actually works as intended.

MohamedAli00949 commented:
I have tested the code with `(\W|^)` and passed all checks 
But after I replaced it with `\b` it didn't pass the python and I didn't discover that at the time.
I have tried to fix the tests but I find it difficult to understand them now so I want any resource (eg: docs, or videos) to understand them easily and fix the tests.

MohamedAli00949 commented:
Now, I have discovered the issue with this code.
The issue is that the word bounding expression selects everything from the beginning to the end as a string and the contains also.
In this sample, you will see that the prompt '...' has been selected as a string not as meta.
![image](https://github.com/highlightjs/highlight.js/assets/76850143/acdd9ee5-1a5f-4d74-81c4-291fc7a765c9)

That's the right show
![image](https://github.com/highlightjs/highlight.js/assets/76850143/db5bf678-f92e-40de-94cc-d4d7d8618063)

I tried to solve it until now it didn't solve it.
My suggested solution is to contact the contains expirations with the beginning expression or go back to use the `(\W|^)` expression.

MohamedAli00949 commented:
Hey @joshgoebel, what is your opinion I understand the test mockups and I think we should put some of them such as `print(test()or'text')` and try that with every string prefix. 
I want your opinion on fixing and testing my changes bugs.

Thanks before any thing

joshgoebel commented:
An example of every string type could work... I assume we'd put that in a test called `strings` or some such if we don't have that already.  As far as the exact regex I'll take another look at the smyself once we have all the test cases.

MohamedAli00949 commented:
Hi @joshgoebel, I have added some markup tests to test my changes and fixed selecting the char before the string prefix as a car such as highlighting the 'plus operator(+)' at the following example `print('kj'+f'kj')`.
Take a look and tell me if you require any other changes.

Many thanks for considering my request.

joshgoebel commented:
In your test you have:

`orbr"https://"`

Is this valid, an `or` followed by a string or invalid?

MohamedAli00949 commented:
> In your test you have:
> 
> `orbr"https://"`
> 
> Is this valid, an `or` followed by a string or invalid?

I think it's invalid, and `orrb"https://"` also.
I want to test `or` with `br` or `rb` prefixes and I have tested them at `or rb"https://"` and `or br"https://"`.
Thanks for the good note

joshgoebel commented:
What is an `rrf` string?

MohamedAli00949 commented:
> What is an `rrf` string?

i think its was for testing if the three letters prefer will be highlighted or highlighting the right string.

joshgoebel commented:
I meant is `rrf` valid python?

MohamedAli00949 commented:
> I meant is `rrf` valid python? Orr `rff`

The `rrf` is not valid python and I have used it to test the regex of begin.
The result does not highlight the prefix and starts highlighting the string from the quotes only.

MohamedAli00949 commented:
Hi @joshgoebel, 
I know that you have done hard work with my PRs and all project PRs and I appreciate that. 
But I don't know what I can do now to merge my PR, fix more issues, and do more on this nice project.

Thanks in advance

joshgoebel commented:
> The result does not highlight any of them and starts highlighting the string from the quotes only.

It's totally OK to include some false positives in the tests (with a comment to explain what purpose they are serving) - but any false positives should still be valid Python code.  We only care about how well we highlight actual real-life valid Python.

MohamedAli00949 commented:
> > The result does not highlight any of them and starts highlighting the string from the quotes only.
> 
> It's totally OK to include some false positives in the tests (with a comment to explain what purpose they are serving) - but any false positives should still be valid Python code. We only care about how well we highlight actual real-life valid Python.

You convinced me, thanks ❤

MohamedAli00949 commented:
> > The result does not highlight any of them and starts highlighting the string from the quotes only.
> 
> It's totally OK to include some false positives in the tests (with a comment to explain what purpose they are serving) - but any false positives should still be valid Python code. We only care about how well we highlight actual real-life valid Python.

I have committed the right tests now, feel free to review and give me your good advice.

MohamedAli00949 commented:
Hi @joshgoebel, 
I know that I have talked more and more to you and I hope to answer my request.
If you have any time please look at the last changes at this request and the last 2 PRs I have pulled before Hecktoperfest finishing.

MohamedAli00949 commented:
Hi @joshgoebel, Any news about reviewing the changes?

MohamedAli00949 commented:
Hi @joshgoebel, If I have removed some of the test commits and collected the testing at one commit would that be best?

joshgoebel commented:
If we just match `or` first - all your tests still pass... do you have additional failing tests you'd like to come up with?

MohamedAli00949 commented:
> If we just match `or` first - all your tests still pass... do you have additional failing tests you'd like to come up with?

Firstly, I am sorry for replying late, I didn't see your comment until now, and thanks for your changes.
Secondly, I think I didn't have any other tests.

joshgoebel commented on src/languages/python.js:
```diff
@@ -214,7 +214,7 @@ export default function(hljs) {
       },
       {
-        begin: /([uU]|[rR])'/,
+        begin: /(([uU]|[rR]))'/,
```
```suggestion
        begin: /([uU]|[rR])'/,
```

joshgoebel commented:
Please add a changelog entry then and we should be good to go.
